### PR TITLE
Create/delete associations when creating/updating a document

### PR DESCRIPTION
**Pull request by tsauerwein**
Closes https://github.com/c2corg/v6_api/issues/274
Closes https://github.com/c2corg/v6_api/issues/180
Closes https://github.com/c2corg/v6_api/issues/212

When creating or updating a document, the associations can now be passed together with the document:

```json
{
    "message": "Update",
    "document": {
        "document_id": 123,
        
        "associations": {
            "users": [
                {"id": 45},
                {"id": 67}
            ],
            "routes": [
                {"document_id": 89}
            ]
        }
    }
}
```

A few remarks:

- For the edit page, a document should now be requested with `...&e=1` to get only the information needed (before there was a parameter `&a=0` to exclude associations, but that name is no longer appropriate).
- For waypoints, parent and child waypoints are returned in two different collections `waypoint_parents` and `waypoint_children`.
- The association types that can be created/updated for a certain document types are listed [here](https://github.com/c2corg/v6_api/blob/associations-edit/c2corg_api/views/validation.py#L296-L303). E.g. for a route you can update associations to other routes and waypoints.

### CODE REVIEW

**Comment by gberaudo** [comment]
For associating (new) images, there was a discussion about directly sending the metadata of the images to associate and have the API:
- create the Image documents;
- create the associations.
@tsauerwein, do you think it is something that could be added easily now or later?

**Comment by tsauerwein** [comment]
@gberaudo, this could be added once the image processing logic is ready. Besides creating the image document and the association, the actual image file would have to be moved from the temporary to the "final" storage.

File: c2corg_api/models/association.py [hunk 1]
```diff
@@ -90,33 +95,71 @@ class AssociationLog(Base):
...
+    if editing_view:
+        edit_types = updatable_associations.get(document.type, set())
+        types_to_include = types_to_include.intersection(edit_types)
+
+    associations = {}
+    if 'waypoints' in types_to_include:
+        waypoint_parents = get_linked_waypoint_parents(document)
+        waypoint_children = get_linked_waypoint_children(document)
+        associations['waypoints'] = waypoint_parents + waypoint_children
+    elif 'waypoint_parents' in types_to_include:
```
Comment (asaunier) [inline c2corg_api/models/association.py]: Not that a big deal but that's a bit not intuitive to get both "waypoint_parents" and "waypoint_children" when ``types_to_include`` only indicates "waypoint_parents" :)

**Comment by asaunier** [comment]
This looks like another huge and impressive PR, thanks :)

> When creating or updating a document, the associations can now be passed together with the document:

If ``associations`` is not passed in the API request, no change occurs in the document's associations?
If it is passed, then all the document's associations are "reset" to its value, right?

> For the edit page, a document should now be request with ...&e=1 to get only the information needed (before there was a parameter &a=0 to exclude associations, but that name is no longer appropriate).

What about documenting this in some README? Ditto for your other remarks?!

File: c2corg_api/models/association.py (hunk 1 above)
Comment (tsauerwein) [inline c2corg_api/models/association.py]: They can only be used together, but I changed it to make it clearer.

**Comment by tsauerwein** [comment]
> If associations is not passed in the API request, no change occurs in the document's associations?
> If it is passed, then all the document's associations are "reset" to its value, right?

If `associations` is not given, no associations are changed. If for example for routes only `"associations": {"waypoints": [...]}` is given, only the the associations to waypoints are updated but not the associations to other routes.

I added some documentation.

**Comment by asaunier** [comment]
Looks good to me!

**Comment by gberaudo** [comment]
+1

**Comment by tsauerwein** [comment]
Thanks for the reviews!

**Pull request merged**